// @tarojs/taro-h5/dist/api/location/index.js
import { temporarilyNotSupport, processOpenApi } from '../../utils/index.js';
export { getLocation } from './getLocation.js';
export { chooseLocation } from './chooseLocation.js';

// 位置
const stopLocationUpdate = /* @__PURE__ */ temporarilyNotSupport('stopLocationUpdate');
const startLocationUpdateBackground = /* @__PURE__ */ temporarilyNotSupport('startLocationUpdateBackground');
const startLocationUpdate = /* @__PURE__ */ temporarilyNotSupport('startLocationUpdate');
const openLocation = /* @__PURE__ */ processOpenApi({
    name: 'openLocation',
    defaultOptions: { scale: 18 }
});
const onLocationChangeError = /* @__PURE__ */ temporarilyNotSupport('onLocationChangeError');
const onLocationChange = /* @__PURE__ */ temporarilyNotSupport('onLocationChange');
const offLocationChangeError = /* @__PURE__ */ temporarilyNotSupport('offLocationChangeError');
const offLocationChange = /* @__PURE__ */ temporarilyNotSupport('offLocationChange');
const choosePoi = /* @__PURE__ */ temporarilyNotSupport('choosePoi');
const getFuzzyLocation = /* @__PURE__ */ temporarilyNotSupport('getFuzzyLocation');

export { choosePoi, getFuzzyLocation, offLocationChange, offLocationChangeError, onLocationChange, onLocationChangeError, openLocation, startLocationUpdate, startLocationUpdateBackground, stopLocationUpdate };
