// @tarojs/components/dist/collection/components/image/image.js
import { h, Host } from '@stencil/core';
import classNames from 'classnames';
// CDN脚本URL
const LEGO_CDN_URL = 'http://ossin.jd.com/swm-plus/h5Tag/tag.js';
// 检查CDN脚本是否已加载
const isLegoScriptLoaded = () => {
  return document.querySelector(`script[src="${LEGO_CDN_URL}"]`) !== null;
};
// 插入CDN脚本
const insertLegoScript = () => {
  if (isLegoScriptLoaded())
    return;
  const script = document.createElement('script');
  script.type = 'module';
  script.src = LEGO_CDN_URL;
  document.head.appendChild(script);
};
// 解析lego协议URL
const parseLegoUrl = (src) => {
  if (!src.startsWith('lego://'))
    return null;
  try {
    // 移除 'lego://' 前缀
    const urlWithoutProtocol = src.substring(7);
    // 分割tagId和参数
    const [tagId, params] = urlWithoutProtocol.split('?');
    // 解析参数
    const text = params ? new URLSearchParams(params).get('text') || '' : '';
    return { tagId, text };
  }
  catch (error) {
    console.warn('Failed to parse lego URL:', src, error);
    return null;
  }
};
export class Image {
  constructor() {
    this.src = undefined;
    this.mode = 'scaleToFill';
    this.lazyLoad = false;
    this.nativeProps = {};
    this.lang = undefined;
    this.aspectFillMode = 'width';
    this.didLoad = false;
  }
  componentDidLoad() {
    // 检查是否为lego模式，如果是则确保CDN脚本已加载
    const legoData = parseLegoUrl(this.src);
    if (legoData !== null) {
      insertLegoScript();
    }
    if (!this.lazyLoad)
      return;
    const lazyImg = new IntersectionObserver(entries => {
      // 异步 api 关系
      if (entries[entries.length - 1].isIntersecting) {
        lazyImg.unobserve(this.imgRef);
        this.didLoad = true;
      }
    }, {
      rootMargin: '300px 0px'
    });
    lazyImg.observe(this.imgRef);
  }
  imageOnLoad() {
    const { width, height, naturalWidth, naturalHeight } = this.imgRef;
    this.onLoad.emit({
      width,
      height
    });
    this.aspectFillMode = naturalWidth > naturalHeight ? 'width' : 'height';
  }
  imageOnError(e) {
    this.onError.emit(e);
  }
  render() {
    const { src, lazyLoad = false, aspectFillMode = 'width', imageOnLoad, imageOnError, nativeProps, didLoad, lang } = this;
    // mode="" 按默认值处理
    const mode = this.mode || 'scaleToFill';
    // 检查是否为lego模式
    const legoData = parseLegoUrl(src);
    const isLegoMode = legoData !== null;
    const cls = classNames({
      'taro-img__widthfix': mode === 'widthFix'
    });
    const imgCls = classNames(`taro-img__mode-${mode.toLowerCase().replace(/\s/g, '')}`, {
      [`taro-img__mode-aspectfill--${aspectFillMode}`]: mode === 'aspectFill'
    });
    // 如果是lego模式，渲染canvas-tag
    if (isLegoMode && legoData) {
      return (h(Host, { class: cls }, h("canvas-tag", Object.assign({ tagId: legoData.tagId, text: legoData.text, lang: lang }, nativeProps))));
    }
    // 普通图片模式
    return (h(Host, { class: cls }, src ? (h("img", Object.assign({ ref: (img) => (this.imgRef = img), class: imgCls, src: lazyLoad && !didLoad ? undefined : src, onLoad: imageOnLoad.bind(this), onError: imageOnError.bind(this) }, nativeProps))) : ''));
  }
  static get is() { return "taro-image-core"; }
  static get originalStyleUrls() {
    return {
      "$": ["./style/index.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["./style/index.css"]
    };
  }
  static get properties() {
    return {
      "src": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "src",
        "reflect": false
      },
      "mode": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "Mode",
          "resolved": "\"aspectFill\" | \"aspectFit\" | \"bottom left\" | \"bottom right\" | \"bottom\" | \"center\" | \"heightFix\" | \"left\" | \"right\" | \"scaleToFill\" | \"top left\" | \"top right\" | \"top\" | \"widthFix\"",
          "references": {
            "Mode": {
              "location": "local"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "mode",
        "reflect": false,
        "defaultValue": "'scaleToFill'"
      },
      "lazyLoad": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "lazy-load",
        "reflect": false,
        "defaultValue": "false"
      },
      "nativeProps": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "{}",
          "resolved": "{}",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "defaultValue": "{}"
      },
      "lang": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "lang",
        "reflect": false
      }
    };
  }
  static get states() {
    return {
      "aspectFillMode": {},
      "didLoad": {}
    };
  }
  static get events() {
    return [{
        "method": "onLoad",
        "name": "load",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": ""
        },
        "complexType": {
          "original": "any",
          "resolved": "any",
          "references": {}
        }
      }, {
        "method": "onError",
        "name": "error",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": ""
        },
        "complexType": {
          "original": "any",
          "resolved": "any",
          "references": {}
        }
      }];
  }
}
